### PR TITLE
Fix CP RNG tracker semantics across dropout and checkpoint loading

### PR DESCRIPTION
**Pull request by ewan0x79**
# What does this PR do ?
Fix RNG tracker semantics for context parallel execution by introducing dedicated CP-only and TPxCP RNG tracker states, routing dropout-related callsites to the correct tracker, and adding backward-compatible checkpoint loading for older RNG state formats.

More specifically, this PR:
- introduces `context-parallel-rng` and `model-and-context-parallel-rng`
- updates CP-sensitive dropout / recompute / graph capture paths to use the correct RNG stream
- overrides TE attention dropout context when sequence parallelism is disabled so TE backends consume the TPxCP RNG tracker consistently
- backfills missing RNG tracker states when loading legacy checkpoints, including the FSDP RNG restore path

Why this change is needed:
- the previous RNG tracker layout did not distinguish CP-only and TPxCP execution
- with context parallelism enabled, that makes dropout / recompute RNG behavior underspecified across some paths
- older checkpoints do not contain the new tracker names, so load-time backfilling is required for backward compatibility

## Issue tracking

Linked issue: N/A (bug fix)

### Pre-checks

- [x] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Testing

Added unit coverage for:
- RNG tracker seed semantics for CP-only and TPxCP states
- embedding dropout behavior under TP=2 / CP=2
- checkpoint recompute preserving CP-aware RNG behavior
- legacy checkpoint loading that is missing the new tracker states
- FSDP RNG state restore compatibility
- TE attention dropout context wiring

### Code review

Feel free to message or comment the [@mcore-oncall](https://github.com/orgs/NVIDIA/teams/mcore-oncall) to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### CODE REVIEW

**Comment by Connor-XY** [comment]
Hello @ewan0x79, thanks for the careful CP RNG tracker fix.

I checked the scope of the diff. This is cross-cutting: it changes CP/TP RNG tracker state names, TE dropout context handling, checkpoint load backfill for older RNG state keys, and the FSDP load path. The breakage being addressed is that CP ranks can replay or restore the wrong dropout RNG state, especially across checkpoint save/load boundaries.

I’m routing this to @wujingyue for the FSDP/CP-RNG coordination path, since this needs one owner to pull together the tensor-parallel, transformer, and checkpointing pieces.

File: megatron/core/tensor_parallel/random.py [hunk 1]
```diff
@@ -89,6 +93,8 @@ def _get_share_storage():
 
 # Default name for the model parallel rng tracker.
 _MODEL_PARALLEL_RNG_TRACKER_NAME = 'model-parallel-rng'
+_CONTEXT_PARALLEL_RNG_TRACKER_NAME = 'context-parallel-rng'
+_MODEL_AND_CONTEXT_PARALLEL_RNG_TRACKER_NAME = 'model-and-context-parallel-rng'
 _EXPERT_PARALLEL_RNG_TRACKER_NAME = 'expert-parallel-rng'
 _DATA_PARALLEL_RNG_TRACKER_NAME = 'data-parallel-rng'
```
Comment (wujingyue) [inline megatron/core/tensor_parallel/random.py:98]: The motivation makes sense to me. I feel like there might be a better structure to maintain these names or RNG streams in the tracker. 

(`_DATA_PARALLEL_RNG_TRACKER_NAME`, `_MODEL_PARALLEL_RNG_TRACKER_NAME`, `_CONTEXT_PARALLEL_RNG_TRACKER_NAME`, `_MODEL_AND_CONTEXT_PARALLEL_RNG_TRACKER_NAME`) is essentially the cartesian product of (whether TP x whether CP). 

(I understood EP is separate because that's for a different portion of the model)

File: megatron/core/tensor_parallel/random.py [hunk 2]
```diff
@@ -213,6 +248,57 @@ def get_data_parallel_rng_tracker_name():
...
+    if _CONTEXT_PARALLEL_RNG_TRACKER_NAME in states:
+        return states
+
+    source_state = fallback_state
+    if source_state is None:
+        source_state = states.get(_DATA_PARALLEL_RNG_TRACKER_NAME)
+    if source_state is None:
+        return states
+
+    states = dict(states)
```
Comment (wujingyue) [inline megatron/core/tensor_parallel/random.py:312]: Looks like a slop.

File: megatron/training/checkpointing.py [hunk 3]
```diff
@@ -1993,10 +1993,19 @@ def load_model_state_dict(module, state_dict, strict: bool):
                 # Check for empty states array
                 if not rng_state['rng_tracker_states']:
                     raise KeyError
+                converted_cuda_rng_state = tensor_parallel.convert_cuda_rng_state(
+                    rng_state['cuda_rng_state'], to_graphable=graph_safe_rng
+                )
```
Comment (wujingyue) [inline megatron/training/checkpointing.py]: This doesn't look related to CP. Does it belong to a different PR?

File: megatron/training/checkpointing.py [hunk 4]
```diff
@@ -1993,10 +1993,19 @@ def load_model_state_dict(module, state_dict, strict: bool):
...
                 rng_tracker_states = {
                     k: tensor_parallel.convert_cuda_rng_state(v, to_graphable=graph_safe_rng)
                     for k, v in rng_state['rng_tracker_states'].items()
                 }
+                rng_tracker_states = tensor_parallel.ensure_context_parallel_rng_tracker_states(
+                    rng_tracker_states, fallback_state=converted_cuda_rng_state
+                )
+                rng_tracker_states = tensor_parallel.ensure_model_and_context_parallel_rng_tracker_states(
+                    rng_tracker_states
+                )
```
Comment (wujingyue) [inline megatron/training/checkpointing.py]: Can this be done after the if-then-else? Note that you are doing the same thing in the if-else branch.

**Review by wujingyue — COMMENTED** [review]
The motivation makes sense. I made some comments but let me also find someone that knows the code better.

File: megatron/training/checkpointing.py (hunk 4 above)
Comment (ewan0x79) [reply megatron/training/checkpointing.py]: I moved the common code outside of the if-else blocks.

File: megatron/core/tensor_parallel/random.py (hunk 2 above)
Comment (ewan0x79) [reply megatron/core/tensor_parallel/random.py:312]: I added a warning message.

File: megatron/training/checkpointing.py (hunk 3 above)
Comment (ewan0x79) [reply megatron/training/checkpointing.py]: This is a fallback for checkpoints without 'context-parallel-rng'; I duplicated it from the TP RNG to handle that case.

File: megatron/core/tensor_parallel/random.py (hunk 1 above)
Comment (ewan0x79) [reply megatron/core/tensor_parallel/random.py:98]: Agreed, the four tracker names do form the Cartesian product of (TP-split, CP-split). A cleaner structure (e.g. an enum or a (use_tp, use_cp) keyed dict) would be nice. However, since tracker names are serialized as checkpoint keys, changing the structure carries backward-compatibility risk and deserves its own PR.

**Comment by ewan0x79** [comment]
In the previous implementation, the seeds were handled as follows:

```python
rng_tracker.add(DP, seed)                                    # seed
rng_tracker.add(TP, seed + 2718 + tp_rank)                   # seed + 2718 + tp_rank
rng_tracker.add(EP, seed + 1024 + 100 * ep_rank + etp_rank)  # seed + 1024 + 100*ep + etp
```

Under the old scheme, the TP seeds form a **contiguous** range
`[seed + 2718, seed + 2718 + TP)` (one consecutive integer per `tp_rank`).

The EP seeds are **not contiguous**: with `etp_rank ∈ [0, ETP)` and `ETP ≪ 100`,
they form sparse clusters of `ETP` values spaced 100 apart, bounded by
`[seed + 1024, seed + 1024 + 100·(EP − 1) + ETP)`.

A collision occurs not merely because these bounding intervals overlap, but
because the equation

```
seed + 2718 + tp_rank = seed + 1024 + 100·ep_rank + etp_rank
⟹ tp_rank = 100·ep_rank + etp_rank − 1694
```

has an integer solution within the valid rank ranges. With `etp_rank = 0`, the
only realistic solution is `ep_rank = 17, tp_rank = 6`, i.e. it requires
`EP ≥ 18` and `TP ≥ 7`. When this holds, the EP seed of the group of GPUs with
`ep_rank = 17` coincides with a TP seed, producing numerically identical seeds
and triggering the duplicate-seed assertion in `CudaRNGStatesTracker.add`.

**Minimal reproduction:** `TP=8, EP=32, ETP=1`. On the GPU
`(tp=6, ep=17, etp=0)`, both the TP and EP seeds equal `seed + 2724`, causing a
crash at startup.

Therefore, in the commit **"Fix CP/TPxCP RNG seed overlap with expert-parallel
seeds by using isolated offsets"**, the seeds for the different parallelism
dimensions are now fully isolated into non-overlapping offset ranges, so the
collision can no longer occur.
